### PR TITLE
Refactoring of code as per mentioned points And Modification of ReadMe

### PR DESCRIPTION
**Pull request by pbhal**
1.) Consolidation of historysegmentcolumn and trendingsegmentcolumn classes to one class

2.) Creation of a helper class for the redundant standard deviation calculation code into one method

3.) Expanded the Validation of the input model in the TrendsCalculator class and removed the validation code from the AbstractTrending Calculator Class

4.)Expanded the test case suite to cover the edge cases.

### CODE REVIEW

File: Library/Core/GlobalZCalculationCriterias/GlobalZCalculationCustomCriteria.cs [hunk 1]
```diff
@@ -34,7 +35,9 @@ internal class GlobalZCalculationCustomCriteria : IGlobalZCalculationCriteria
 
             double countelements = (count * 1.0);
             mean = sumHistorySegmentColumns / countelements;
-            standardDeviation = CalculateStandardDeviation(values, mean);
+
+            CalculationHelper calculationHelper = new CalculationHelper();
```
Comment (pwalecha) [inline Library/Core/GlobalZCalculationCriterias/GlobalZCalculationCustomCriteria.cs]: this can be the static class if the intent is to create like a helper.

File: Library/Core/GlobalZCalculationCriterias/GlobalZCalculationZMeanCriteria.cs [hunk 2]
```diff
@@ -38,7 +39,9 @@ internal class GlobalZCalculationZMeanCriteria : IGlobalZCalculationCriteria
 
             double countelements = (count * 1.0);
             mean = sumHistorySegmentColumns / countelements;
-            standardDeviation = CalculateStandardDeviation(values, mean);
+
+            CalculationHelper calculationHelper = new CalculationHelper();
```
Comment (pwalecha) [inline Library/Core/GlobalZCalculationCriterias/GlobalZCalculationZMeanCriteria.cs]: Make use of the helper class

File: Library/Helper/CalculationHelper.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace TrendsCalculator.Library.Helper
+{
+    internal class CalculationHelper
+    {
+        internal double CalculateStandardDeviation(List<int> values, double mean)
```
Comment (pwalecha) [inline Library/Helper/CalculationHelper.cs]: make this method as static and use directly.

File: Test/TrendsCalculatorTest.cs [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+namespace TrendsCalculator.Test
+{
+    [TestClass]
+    public class TrendsCalculatorTest
+    {
+        [TestMethod]
+        public void TrendsCalculatorTest_CustomCriteria_Suite1()
+        {
+            var trendingCalculator = new Library.TrendsCalculator(TrendCalculationStrategy.Custom);
+            var inputData = DataPreparator.PrepareData_Suite1();
```
Comment (pwalecha) [inline Test/TrendsCalculatorTest.cs:17]: Instead of naming suite1 in the name: Give an intuitive name of what the test case is about. It is Ok if the method name gets a little long.. E.g.You need not write "TrendsCalculatorTest" as this is implicit.
E.g. Test_CustomCritera_BlankRecords_ShouldGiveException
Test_CustomCriteria_InsufficientDemand_ShouldMakeElementMostTrending

**Review by pwalecha — APPROVED** [review]

File: README.md [hunk 5]
```diff
@@ -12,36 +12,40 @@ To install this library, please download the latest version of  [NuGet Package](
 ## How to use it
 The library takes four input parameters:-
 
-1. Enum of the type **TrendsCalculationStrategy** which has two predefined enumeration values signifying the type of stratregy required in filtering of results:
+1. Enum of the type **TrendsCalculationStrategy** which has three predefined enumeration values signifying the type of stratregy required in filtering of results:
    1. **ZMean**:- This Enum evaluates and return trending models based on the Z mean value (threshold value is z mean value).
    2. **Custom**:- This Enum gives the the entire list sorted in order of trends evaluated (sorted by global Z, local Z).
    3. **DemandSupply**:- This Enum evaluates and return trending models based on the Demand and Supply. If there is a Demand for a particular entity over a certain period and
```
Comment (pwalecha) [inline README.md]: DemandSupply is not a part of enum anymore - It is included in a separate method.

File: README.md [hunk 6]
```diff
@@ -12,36 +12,40 @@ To install this library, please download the latest version of  [NuGet Package](
...
-    Each of the model type T, should inherit from the TInterface of the NuGet package.
+    Each of the model type T, should extend from either the TModel or the TDemandSupplyModel of the NuGet package.
     
     The algorithm takes into account the virtual windowsPeriod, lets say the algorithm is run for the windowPeriod of 6 months in the month of June 2020 and the data for June is not yet complete, so in order to compensate for that , the data can be provided for December 2019 to June 2020.If not using the virtual windowPeriod, the data must be provided for January 2020 to June 2020.
 
-First, an object of the TrendingDataCalculator class needs to be instantiated with the constructor which takes Enum as input defining the strategy.The example of the code is :-
+First, an object of the TrendsCalculator class needs to be instantiated with the constructor which takes Enum as input defining the strategy.The example of the code is :-
 ```
-var trendingDataCalculator = new TrendingDataCalculator<YourTrendingModelType>(TrendCalculationStrategy.ZMean);
+var trendingCalculator = new TrendsCalculator.Library.TrendsCalculator(TrendCalculationStrategy.DemandSupply);
```
Comment (pwalecha) [inline README.md]: not applicable

**Review by pwalecha — APPROVED** [review]
